sql: fix INSERT INTO t DEFAULT VALUES in presence of schema changes

jordanlewis opened the pull request:
Previously, `INSERT INTO t DEFAULT VALUES` would fail if there were
outstanding column schema changes.

Closes #29478.

Release note (bug fix): correct behavior of `INSERT INTO t DEFAULT
VALUES` when there are active schema changes.

jordanlewis committed:
sql: fix INSERT INTO t DEFAULT VALUES in presence of schema changes

Previously, `INSERT INTO t DEFAULT VALUES` would fail if there were
outstanding column schema changes.

Release note (bug fix): correct behavior of `INSERT INTO t DEFAULT
VALUES` when there are active schema changes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/29479)
<!-- Reviewable:end -->

jordanlewis commented:
The test looks like it changed a lot, but I just added subtests to it to make the output clearer and added a new test case for `INSERT/UPSERT INTO t DEFAULT VALUES`.

vivekmenezes reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29479)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/insert.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29479#-LLX6sD7A4pKp3iIt8tF:-LLX6sD7A4pKp3iIt8tG:b6gg3nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdcb5e9e96bb7b048b35e6cd95ac53a30f580329/pkg/sql/insert.go#L120)):*
> ```Go
> 	if n.DefaultValues() {
> 		// No target column, select all columns in the table, including
> 		// hidden columns; these may have defaults too.
> ```

Wow, the comment talks about hidden columns :-)


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29479)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/insert.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29479#-LLX6sD7A4pKp3iIt8tF:-LLX8eBYEAo5Wp69IzaS:b-wgrude) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdcb5e9e96bb7b048b35e6cd95ac53a30f580329/pkg/sql/insert.go#L120)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Wow, the comment talks about hidden columns :-)
</blockquote></details>

Heh, yeah, except it's talking about the *real* kind of hidden columns (`rowid` is the only example) and not mutation columns...

This whole situation really does need a refactor. I was thinking an interface as simple as `desc.GetColumns(mutation level)` could be sufficient, and maybe add linting to remove direct usage of `desc.Columns`. I doubt it's quite this simple, but one can only dream... I feel like I've done this exact kind of fix 3-4 times in the last year.


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (dismissed):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29479)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/insert.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29479#-LLX6sD7A4pKp3iIt8tF:-LLX96xdEO8ynat2a6jh:b-tdaq8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdcb5e9e96bb7b048b35e6cd95ac53a30f580329/pkg/sql/insert.go#L120)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Heh, yeah, except it's talking about the *real* kind of hidden columns (`rowid` is the only example) and not mutation columns...

This whole situation really does need a refactor. I was thinking an interface as simple as `desc.GetColumns(mutation level)` could be sufficient, and maybe add linting to remove direct usage of `desc.Columns`. I doubt it's quite this simple, but one can only dream... I feel like I've done this exact kind of fix 3-4 times in the last year.
</blockquote></details>

That is what creating a new ExtendedTableDescriptor is meant to do. We should be computing all these once and using them.
___
*[pkg/sql/insert.go, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29479#-LLX8VtqAZNByld56Gly:-LLX8Vtr8c_Myqt1nCr8:bpfwnm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdcb5e9e96bb7b048b35e6cd95ac53a30f580329/pkg/sql/insert.go#L125)):*
> ```Go
> 		for i := range desc.Mutations {
> 			if c := desc.Mutations[i].GetColumn(); c != nil && desc.Mutations[i].State == sqlbase.DescriptorMutation_DELETE_AND_WRITE_ONLY {
> 				insertCols = append(insertCols, *c)
> ```

I worry about using the underlying array in a shared slice. The descriptor is supposed to be immutable. I'd suggest doing 

```
if len(desc.Mutations) > 0 {

      insertCols := make([]ColumnDescriptor, 0, len(desc.Columns) + len(dec.Mutations)
      blah

}

```


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29479)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/insert.go, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29479#-LLX8VtqAZNByld56Gly:-LLXKNPO9DMpEH5GJqbX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdcb5e9e96bb7b048b35e6cd95ac53a30f580329/pkg/sql/insert.go#L125)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I worry about using the underlying array in a shared slice. The descriptor is supposed to be immutable. I'd suggest doing 

```
if len(desc.Mutations) > 0 {

      insertCols := make([]ColumnDescriptor, 0, len(desc.Columns) + len(dec.Mutations)
      blah

}

```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
bors r+
